File: CourseProjects/CourseProjects/JavaEclipseProjects/UnitTestingExercises/src/exercise/tdd/videostore/start/Customer.java
/*
 * This example adapted from 'Refactoring' by Martin Fowler
 *  
 */
package exercise.tdd.videostore.start;

import static exercise.tdd.videostore.start.PriceCode.*;
import java.util.ArrayList;
import java.util.List;

public class Customer {
    private String name;
    private List<Rental> rentals = new ArrayList<Rental>();
    public Customer(String name) {
        this.name = name;
    }
    public String getName() {
        return name;
    }
    public void addRental(Rental arg) {
        rentals.add(arg);
    }
    public String statement() {
        double totalAmount = 0;
        int frequentRenterPoints = 0;
        StringBuilder result = new StringBuilder();
        addHeaderLine(result);
        for (Rental each : rentals) {
            double thisAmount = calcAmount(each);
            // add frequent renter points
            frequentRenterPoints++;
            // add bonus for a two day new release rental
            if ((each.getMovie().getPriceCode() == NEW_RELEASE)
                    && each.getDaysRented() > 1)
                frequentRenterPoints++;
            addStatementBody(result, each, thisAmount);
            //add cost of this rental to total cost
            totalAmount += thisAmount;
        }
        addFooterLines(totalAmount, frequentRenterPoints, result);
        return result.toString();
    }
	private void addFooterLines(double totalAmount, int frequentRenterPoints,
			StringBuilder result) {
		result.append("Amount owed is ");
        result.append(totalAmount);
        result.append("\n");
        result.append("You earned ");
        result.append(frequentRenterPoints);
        result.append(" frequent renter points\n");
	}
	private void addStatementBody(StringBuilder result, Rental each, double thisAmount) {
		result.append("\t");
		result.append(each.getDaysRented());
		result.append("\t");
		result.append(each.getMovie().getTitle());
		result.append("\t");
		result.append(thisAmount);
		result.append("\n");
	}
	private void addHeaderLine(StringBuilder result) {
		result.append("\nRental Record for ");
        result.append(getName());
        result.append("\n");
	}
	private double calcAmount(Rental each) {
		double thisAmount = 0;
		//determine amounts for each line
		switch (each.getMovie().getPriceCode()) {
		case REGULAR:
		    thisAmount += 2;
		    if (each.getDaysRented() > 2)
		        thisAmount += (each.getDaysRented() - 2) * 1.5;
		    break;
		case NEW_RELEASE:
		    thisAmount += each.getDaysRented() * 3;
		    break;
		case CHILDRENS:
		    thisAmount += 1.5;
		    if (each.getDaysRented() > 3)
		        thisAmount += (each.getDaysRented() - 3) * 1.5;
		    break;
		}
		return thisAmount;
	}
}

File: CourseProjects/CourseProjects/VisualStudioProjects/Agile/RefactoringChallenge/VideostoreRefactoringStart/Customer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace VideostoreRefactoringStart {
    public class Customer {
        public Customer(String name) {
            this.name = name;
            rentals = new List<Rental>();
        }
        public String Name {
            get { return name; }
        }
        public void AddRental(Rental arg) {
            rentals.Add(arg);
        }
        public string Statement() {
            double totalAmount = 0;
            int frequentRenterPoints = 0;
            StringBuilder result = new StringBuilder();
            //add header lines
            result.Append("\nRental Record for ");
            result.Append(Name);
            result.Append("\n");
            foreach (Rental each in rentals) {
                double thisAmount = 0;
                //determine amounts for each line
                switch (each.Movie.PriceCode) {
                case PriceCode.REGULAR:
                    thisAmount += 2;
                    if (each.DaysRented > 2)
                        thisAmount += (each.DaysRented - 2) * 1.5;
                    break;
                case PriceCode.NEW_RELEASE:
                    thisAmount += each.DaysRented * 3;
                    break;
                case PriceCode.CHILDRENS:
                    thisAmount += 1.5;
                    if (each.DaysRented > 3)
                        thisAmount += (each.DaysRented - 3) * 1.5;
                    break;
                }
                // add frequent renter points
                frequentRenterPoints++;
                // add bonus for a two day new release rental
                if ((each.Movie.PriceCode == PriceCode.NEW_RELEASE)
                        && each.DaysRented > 1)
                    frequentRenterPoints++;
                //show figures for this rental
                result.Append("\t");
                result.Append(each.DaysRented);
                result.Append("\t");
                result.Append(each.Movie.Title);
                result.Append("\t");
                result.Append(thisAmount);
                result.Append("\n");
                //add cost of this rental to total cost
                totalAmount += thisAmount;
            }
            //add footer lines
            result.Append("Amount owed is ");
            result.Append(totalAmount);
            result.Append("\n");
            result.Append("You earned ");
            result.Append(frequentRenterPoints);
            result.Append(" frequent renter points\n");
            return result.ToString();
        }
        private String name;
        private List<Rental> rentals;
    }
}
